### DISCUSSION TITLE
Writing to multiple output pixels using vips_image_generate

### DISCUSSION THREAD

**Discussion by DDoS** [libvips/libvips#3878] [Q&A]
For an operation I'm trying to implement I need to write multiple pixels for each input pixel (error diffusion). I saw mentioned in the doc that you can change the size of the input region, but can that also be done for the output region? Is there a way to make sure that the `out_region` is always NxM pixels bigger than the input region?

Really I just want to implement "Floyd–Steinberg" dithering, and I'm not sure who to correctly write to neighbouring pixels.

**Comment by jcupitt** [comment]
Hi @DDoS,

I think enlarging the output would mean that a pixel could be written more than once at overlaps, which wouldn't work.

You need a dither algorithm with limited context, ie. each output pixel can be computed from a bounded set of input pixels, ie. you have some term that stops errors being diffused too far. Now you can just enlarge the input by the appropriate amount and compute the output area.

libvips has a built-in dither, I guess you saw? If you save a GIF or an 8-bit PNG it'll be dithered for you:

```
vips copy nina.jpg x.png[palette,bitdepth=4]
```

To make a 16-colour dithered PNG:

![x](https://github.com/libvips/libvips/assets/580843/feebd8b1-b1d3-4f1d-b4e3-9ce485ec574e)

**Comment by DDoS** [comment]
Thanks for the quick reply. I hadn't noticed the dither support, looks like it's a private API? I need to use a custom palette and in Lab space.

Anyway I decided to ditch `vips_image_generate` and instead do the dithering line-by-line with `vips_region_prepare`, using a memory copy of the input image. I need to write the final indexed image to an external buffer anyway, and it's small enough that memory usage isn't an issue. I have some out-of-gamut issues, but otherwise this seems to work, so I'll mark the issue are resolved.
